### PR TITLE
feat: ability to add new editor themes to Fiddle

### PR DESCRIPTION
**Pull request by cvaldez98**
Users have had the ability to create custom themes but this makes it easier for them to bring editor-only themes onto Fiddle. 
We now simply use whatever theme the user is currently using and replace the `editor` under the `FiddleTheme` interface with the JSON file the user puts in.  This makes it easy for users to use a VS Code theme if they want to. 
To test this I went to `https://bitwiser.in/monaco-themes/` and then copied and pasted the JSON onto a file. I then selected that file from Fiddle and I had a cool new theme set up. 

image

Editor with the *monoindustrial* editor theme
image

### CODE REVIEW

**Comment by erickzhao** [comment]
Just an idea: could we let people select their VSCode extensions folder (with the dialog's `defaultPath` set to `.vscode/extensions/`, and automatically add all those themes to `getAvailableThemes()` automatically? And have this enabled/disabled via checkbox.

File: src/renderer/components/settings-general-appearance.tsx [hunk 1]
```diff
@@ -193,13 +226,36 @@ export class AppearanceSettings extends React.Component<
...
   }
+
+  private async promptForTheme(currWindow: Electron.BrowserWindow, defaultTheme: LoadedFiddleTheme) {
+    const filePicked = await remote.dialog.showOpenDialog(currWindow, {
+      title: 'Pick a VSCode editor theme file',
+      properties: ['openFile'],
+      filters: [ { name: 'JSON', extensions: ['json']}]
+    });
+    if (filePicked === undefined || filePicked.length === 0) {
+      return Error('Prompt cancelled by user');
```
[the remark this answers is not in the document]
Comment (cvaldez98) [reply src/renderer/components/settings-general-appearance.tsx]: The error will be catched by the try catch above it and will simply `console.warn('error', error)` so it won't crash or do anything terrible

File: src/renderer/components/settings-general-appearance.tsx [hunk 2]
```diff
@@ -193,13 +226,36 @@ export class AppearanceSettings extends React.Component<
...
+            text='Add a VSCode Editor theme'
+            icon='duplicate'
+          />
         </Callout>
       </div>
     );
   }
+
+  private async promptForTheme(currWindow: Electron.BrowserWindow, defaultTheme: LoadedFiddleTheme) {
+    const filePicked = await remote.dialog.showOpenDialog(currWindow, {
```
[the remark this answers is not in the document]
Comment (cvaldez98) [reply src/renderer/components/settings-general-appearance.tsx]: Sorry I'm not sure I understand, could you elaborate more on this? I do need the current window in order to call `showOpenDialog` since it is a required parameter.

[the remark this answers is not in the document]
Comment (erickzhao) [reply src/renderer/components/settings-general-appearance.tsx]: image
https://electronjs.org/docs/api/dialog

File: src/renderer/components/settings-general-appearance.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (erickzhao) [reply src/renderer/components/settings-general-appearance.tsx]: Do we need to warn the user if they cancel the action manually? It might be cleaner to just return nothing.

File: src/renderer/components/settings-general-appearance.tsx [hunk 3]
```diff
@@ -193,13 +226,36 @@ export class AppearanceSettings extends React.Component<
             </a>. The easiest way to get started is to clone one of the two existing
             themes and to add your own colors.
           </p>
+          <p>
+            Additionally, if you wish to import an editor theme from VSCode, just pick your JSON file and Fiddle will do the rest.
+          </p>
           <Button
```
[the remark this answers is not in the document]
Comment (cvaldez98) [reply src/renderer/components/settings-general-appearance.tsx]: That kinda makes it look more cluttered :( 
image

File: src/renderer/components/settings-general-appearance.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (cvaldez98) [reply src/renderer/components/settings-general-appearance.tsx]: Ahhh VSCode did me dirty, 😄

File: src/renderer/components/settings-general-appearance.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (cvaldez98) [reply src/renderer/components/settings-general-appearance.tsx]: Trueeeee

File: tests/renderer/components/__snapshots__/settings-general-appearance-spec.tsx.snap [hunk 4]
```diff
@@ -63,11 +63,19 @@ exports[`AppearanceSettings component renders 1`] = `
       </a>
       . The easiest way to get started is to clone one of the two existing themes and to add your own colors.
     </p>
+    <p>
+      Additionally, if you wish to import an editor theme from VSCode, just pick your JSON file and Fiddle will do the rest.
```
Comment (erickzhao) [inline tests/renderer/components/__snapshots__/settings-general-appearance-spec.tsx.snap]: In its current state, we should probably mention that this is a feature for importing from Monaco rather than a VSCode JSON file, since the latter can't be imported directly.

File: src/renderer/components/settings-general-appearance.tsx [hunk 5]
```diff
@@ -193,13 +224,41 @@ export class AppearanceSettings extends React.Component<
...
   }
+
+  private async promptForTheme(defaultTheme: LoadedFiddleTheme) {
+    const filePicked = await remote.dialog.showOpenDialog({
+      title: 'Pick a VSCode editor theme file',
+      properties: ['openFile'],
+      filters: [ { name: 'JSON', extensions: ['json']}]
+    });
+    if (filePicked === undefined || filePicked.length === 0) {
+      return Error('Prompt cancelled by user');
```
Comment (erickzhao) [inline src/renderer/components/settings-general-appearance.tsx]: nit: I think returning an `Error` here and returning `false` from the parent function isn't very idiomatic. Could instead simply return early or return `null` and check for `!name` in the caller.

File: src/renderer/components/settings-general-appearance.tsx [hunk 6]
```diff
@@ -193,13 +224,41 @@ export class AppearanceSettings extends React.Component<
             </a>. The easiest way to get started is to clone one of the two existing
             themes and to add your own colors.
           </p>
+          <p>
+            Additionally, if you wish to import an editor theme from VSCode, just pick your JSON file and Fiddle will do the rest.
```
Comment (erickzhao) [inline src/renderer/components/settings-general-appearance.tsx]:
```suggestion
            Additionally, if you wish to import a Monaco Editor theme, just pick your JSON file and Fiddle will do the rest.
```

**Comment by erickzhao** [comment]
@cvaldez98 Reran coveralls, your PR should be good now, but it seems like b3743a4 reverted the `VSCode -> Monaco` changes in the text. Mind reverting that commit back?

File: src/renderer/components/settings-general-appearance.tsx [hunk 7]
```diff
@@ -193,13 +241,70 @@ export class AppearanceSettings extends React.Component<
...
+  private async findThemeDirectories(currentDir: string, visited: Set<string> = new Set(), found: Set<string> = new Set()) {
+    if (visited.has(currentDir)) return found;
+    await Promise.all(childDirs(currentDir).map(async (child) => {
+      child = path.resolve(currentDir, child);
+      if (isTheme(child)) found.add(child);
+      await this.findThemeDirectories(child, visited, found);
+    }));
+    return found;
+
+    function childDirs(dir: string) {
```
Comment (MarshallOfSound) [inline src/renderer/components/settings-general-appearance.tsx]: Can this be `async` and use `await fsType.readdir`

File: src/renderer/components/settings-general-appearance.tsx [hunk 8]
```diff
@@ -193,13 +241,70 @@ export class AppearanceSettings extends React.Component<
...
+      return fsType.readdirSync(dir).filter((f) => fsType.lstatSync(path.resolve(dir, f)).isDirectory());
+    }
+    // path.basename
+    function isTheme(dir: string) {
```
Comment (MarshallOfSound) [inline src/renderer/components/settings-general-appearance.tsx]: Same here, should be `async` and use `await fsType.lstat`

File: src/renderer/components/settings-general-appearance.tsx [hunk 9]
```diff
@@ -193,13 +224,47 @@ export class AppearanceSettings extends React.Component<
...
+            text='Add a Monaco Editor theme'
+            icon='duplicate'
+          />
         </Callout>
       </div>
     );
   }
+
+  private async promptForTheme(defaultTheme: LoadedFiddleTheme) {
+    const filePicked = await remote.dialog.showOpenDialog({
```
Comment (felixrieseberg) [inline src/renderer/components/settings-general-appearance.tsx]: The `remote` object might be going away soon (possibly), so we should try to avoid using it in new code. Could you find a way to do this without the `dialog` module at all? Maybe just with a native HTML5 file dialog? From the looks of it, the HTML5 version would do everything you need: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file

File: src/renderer/components/settings-general-appearance.tsx [hunk 10]
```diff
@@ -193,13 +224,47 @@ export class AppearanceSettings extends React.Component<
             </a>. The easiest way to get started is to clone one of the two existing
             themes and to add your own colors.
           </p>
+          <p>
+            Additionally, if you wish to import a Monaco Editor theme, just pick your JSON file and Fiddle will do the rest.
```
Comment (felixrieseberg) [inline src/renderer/components/settings-general-appearance.tsx]:
```suggestion
            Additionally, if you wish to import a Monaco Editor theme, pick your JSON file and Fiddle will do the rest.
```

I'm a little worried about how optimistic that line is given how likely this feature is to break in the future 🙈 Can we rephrase it to speak about an attempt to import it?

File: src/renderer/components/settings-general-appearance.tsx [hunk 11]
```diff
@@ -139,6 +141,35 @@ export class AppearanceSettings extends React.Component<
     }
   }
 
+  /**
+   * Creates a new template that takes in Monaco editor JSON theme only.
+   * @returns {Promise<boolean>}
+   * @memberof AppearanceSettings
+   */
+  public async createNewThemeFromMonaco(): Promise<boolean> {
+    // fetch current theme
```
Comment (felixrieseberg) [inline src/renderer/components/settings-general-appearance.tsx]: Can you add some line breaks in this method to break it up into logical pieces?

File: src/renderer/components/settings-general-appearance.tsx [hunk 12]
```diff
@@ -1,11 +1,12 @@
 import { Button, Callout, FormGroup, MenuItem } from '@blueprintjs/core';
 import { ItemPredicate, ItemRenderer, Select } from '@blueprintjs/select';
-import { shell } from 'electron';
+import { remote, shell } from 'electron';
 import * as fsType from 'fs-extra';
 import { observer } from 'mobx-react';
 import * as path from 'path';
 import * as React from 'react';
 
+import * as MonacoType from 'monaco-editor';
```
Comment (felixrieseberg) [inline src/renderer/components/settings-general-appearance.tsx]: Move that line up into the block with the other module imports

**Comment by ckerr** [comment]
pinging @felixrieseberg and @MarshallOfSound for a revisit

File: src/renderer/components/dialog-add-monaco-theme.tsx [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    const defaultTheme = !!currentTheme ? await getTheme(currentTheme) : defaultDark;
+    if (!file) return;
+
+    try {
+      const editor = fsType.readJSONSync(file.path);
+      if (!editor.base && !editor.rules) return; // has to have these attributes
+      defaultTheme.editor = editor as Partial<MonacoType.editor.IStandaloneThemeData>;
+      const newTheme = defaultTheme;
+      const name = editor.name ? editor.name : file.name;
+      this.createNewThemeFromMonaco(name, newTheme);
```
[the remark this answers is not in the document]
Comment (cvaldez98) [reply src/renderer/components/dialog-add-monaco-theme.tsx]: True!

File: src/renderer/components/dialog-add-monaco-theme.tsx [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+
+  /**
+   * Handles the submission of the dialog
+   *
+   * @returns {Promise<void>}
+   */
+  public async onSubmit(): Promise<void> {
+    const { file } = this.state;
+    const currentTheme = this.props.appState.theme;
+    const defaultTheme = !!currentTheme ? await getTheme(currentTheme) : defaultDark;
```
[the remark this answers is not in the document]
Comment (cvaldez98) [reply src/renderer/components/dialog-add-monaco-theme.tsx]: Done, I think it was more of it just being more readable at the time I wrote this

File: src/renderer/components/dialog-add-monaco-theme.tsx [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+    this.reset();
+  }
+
+  public render() {
+    const { isMonacoVersionDialogShowing } = this.props.appState;
+    // const inputProps = { webkitdirectory: 'true' };
+    const inputProps = { accept: '.json' };
+    const { file } = this.state;
+
+    const text = file && file.path ? file.path : `Select the JSON Monaco file...`;
```
[the remark this answers is not in the document]
Comment (cvaldez98) [reply src/renderer/components/dialog-add-monaco-theme.tsx]: I just changed it to monaco file 🤷‍♂️open to ideas though

File: src/renderer/components/dialog-add-monaco-theme.tsx [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+   */
+  public async onSubmit(): Promise<void> {
+    const { file } = this.state;
+    const currentTheme = this.props.appState.theme;
+    const defaultTheme = !!currentTheme ? await getTheme(currentTheme) : defaultDark;
+    if (!file) return;
+
+    try {
+      const editor = fsType.readJSONSync(file.path);
+      if (!editor.base && !editor.rules) return; // has to have these attributes
```
[the remark this answers is not in the document]
Comment (cvaldez98) [reply src/renderer/components/dialog-add-monaco-theme.tsx]: Done, it should know show a warning prompt.

**Comment by felixrieseberg** [comment]
@cvaldez98 Thank you so much for bearing with all these recommendations. I think the PR introduces an interesting new feature and improves the code base, so I'll merge it in.

@erickzhao @MarshallOfSound If you have outstanding requests, feel free to add them to this PR and I'll make sure that all additional concerns are addressed before we release a new version. Thank you for your reviews!
